### PR TITLE
Benchmark samples for WorkManager

### PR DESCRIPTION
**Pull request by dlam**

### CODE REVIEW

File: WorkManagerSample/app/src/androidTest/java/com/example/background/ImageOperationsTest.kt [hunk 1]
```diff
@@ -35,10 +35,10 @@ import androidx.work.Configuration
 import androidx.work.WorkManager
 import androidx.work.testing.SynchronousExecutor
 import androidx.work.testing.WorkManagerTestInitHelper
-import com.example.background.Constants.KEY_IMAGE_URI
```
Comment (ChrisCraik) [inline WorkManagerSample/app/src/androidTest/java/com/example/background/ImageOperationsTest.kt]: keep the packages the same, so we can merge back easily

File: WorkManagerSample/benchmark/proguard-rules.pro [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# Add project specific ProGuard rules here.
```
Comment (ChrisCraik) [inline WorkManagerSample/benchmark/proguard-rules.pro]: remove this file

File: WorkManagerSample/benchmark/src/androidTest/AndroidManifest.xml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools"
+    package="com.example.benchmark">
+
```
Comment (ChrisCraik) [inline WorkManagerSample/benchmark/src/androidTest/AndroidManifest.xml]: copy comment from androidx:

File: WorkManagerSample/benchmark/src/androidTest/java/com/example/benchmark/WorkerBenchmark.kt [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+@RunWith(AndroidJUnit4::class)
+@SmallTest
+class WorkerBenchmark {
+
+    companion object {
+        private const val JETPACK = "${BaseFilterWorker.ASSET_PREFIX}images/jetpack.png"
+    }
+
+    private lateinit var mContext: Context
+    private lateinit var mTargetContext: Context
```
Comment (ChrisCraik) [inline WorkManagerSample/benchmark/src/androidTest/java/com/example/benchmark/WorkerBenchmark.kt]: should just need targetContext, and don't need lateinit

File: WorkManagerSample/benchmark/src/androidTest/java/com/example/benchmark/WorkerBenchmark.kt [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+@RunWith(AndroidJUnit4::class)
+@SmallTest
+class WorkerBenchmark {
+
+    companion object {
+        private const val JETPACK = "${BaseFilterWorker.ASSET_PREFIX}images/jetpack.png"
+    }
+
+    private lateinit var mContext: Context
```
Comment (ChrisCraik) [inline WorkManagerSample/benchmark/src/androidTest/java/com/example/benchmark/WorkerBenchmark.kt]: don't use mFoo prefix in kotlin, only in .java

File: WorkManagerSample/benchmark/src/androidTest/java/com/example/benchmark/WorkerBenchmark.kt [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+            worker.applyFilter(BitmapFactory.decodeStream(inputStream))
+            inputStream?.close()
+        }
+    }
+
+    private fun <T : ListenableWorker> createWorker(
+            createWorkerBlock: (appContext: Context, workerClassName: String,
+                                workerParameters: WorkerParameters) -> T
+    ): T {
+        lateinit var worker: T
```
Comment (ChrisCraik) [inline WorkManagerSample/benchmark/src/androidTest/java/com/example/benchmark/WorkerBenchmark.kt]: add a comment about the only way to instantiate a worker is to enqueue work into workmanager

File: WorkManagerSample/benchmark/src/androidTest/java/com/example/benchmark/WorkerBenchmark.kt [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+                .build()
+
+        WorkManagerTestInitHelper.initializeTestWorkManager(mTargetContext, configuration)
+        val workManager = WorkManager.getInstance()
+
+        // Enqueue a WorkRequest in a blocking way to guarantee WorkManager has instantiated the worker.
+        val workRequest = OneTimeWorkRequestBuilder<WaterColorFilterWorker>()
+                .setInputData(workDataOf(Constants.KEY_IMAGE_URI to JETPACK))
+                .build()
+        runBlocking {
```
Comment (ChrisCraik) [inline WorkManagerSample/benchmark/src/androidTest/java/com/example/benchmark/WorkerBenchmark.kt]: why runBlocking here?

File: WorkManagerSample/benchmark/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -0,0 +1,3 @@
+<resources>
```
Comment (ChrisCraik) [inline WorkManagerSample/benchmark/src/main/res/values/strings.xml]: remove file

File: WorkManagerSample/lib/proguard-rules.pro [hunk 9]
```diff
@@ -0,0 +1,21 @@
+# Add project specific ProGuard rules here.
```
Comment (ChrisCraik) [inline WorkManagerSample/lib/proguard-rules.pro]: remove file

File: WorkManagerSample/lib/src/main/res/values/strings.xml [hunk 10]
```diff
@@ -0,0 +1,3 @@
+<resources>
```
Comment (ChrisCraik) [inline WorkManagerSample/lib/src/main/res/values/strings.xml]: remove file

File: gradlew [hunk 11]
```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/env sh
```
Comment (ChrisCraik) [inline gradlew]: why is this added?

Comment (ChrisCraik) [reply gradlew]: I wouldn't expect any changes to the gradle/ directory, including those above/below

File: WorkManagerSample/app/src/androidTest/java/com/example/background/ImageOperationsTest.kt (hunk 1 above)
Comment (dlam) [reply WorkManagerSample/app/src/androidTest/java/com/example/background/ImageOperationsTest.kt]: :+1:

File: WorkManagerSample/benchmark/proguard-rules.pro (hunk 2 above)
Comment (dlam) [reply WorkManagerSample/benchmark/proguard-rules.pro]: :+1:

File: WorkManagerSample/benchmark/src/androidTest/AndroidManifest.xml (hunk 3 above)
Comment (dlam) [reply WorkManagerSample/benchmark/src/androidTest/AndroidManifest.xml]: :+1:

File: WorkManagerSample/benchmark/src/androidTest/java/com/example/benchmark/WorkerBenchmark.kt (hunk 4 above)
Comment (dlam) [reply WorkManagerSample/benchmark/src/androidTest/java/com/example/benchmark/WorkerBenchmark.kt]: :+1:

File: WorkManagerSample/benchmark/src/androidTest/java/com/example/benchmark/WorkerBenchmark.kt (hunk 5 above)
Comment (dlam) [reply WorkManagerSample/benchmark/src/androidTest/java/com/example/benchmark/WorkerBenchmark.kt]: :+1:

File: WorkManagerSample/benchmark/src/main/res/values/strings.xml (hunk 8 above)
Comment (dlam) [reply WorkManagerSample/benchmark/src/main/res/values/strings.xml]: :+1:

File: WorkManagerSample/lib/proguard-rules.pro (hunk 9 above)
Comment (dlam) [reply WorkManagerSample/lib/proguard-rules.pro]: :+1:

File: WorkManagerSample/lib/src/main/res/values/strings.xml (hunk 10 above)
Comment (dlam) [reply WorkManagerSample/lib/src/main/res/values/strings.xml]: :+1:

File: gradlew (hunk 11 above)
Comment (dlam) [reply gradlew]: :+1:

File: WorkManagerSample/benchmark/src/androidTest/java/com/example/benchmark/WorkerBenchmark.kt (hunk 6 above)
Comment (dlam) [reply WorkManagerSample/benchmark/src/androidTest/java/com/example/benchmark/WorkerBenchmark.kt]: :+1:

File: WorkManagerSample/benchmark/src/androidTest/java/com/example/benchmark/WorkerBenchmark.kt (hunk 7 above)
Comment (dlam) [reply WorkManagerSample/benchmark/src/androidTest/java/com/example/benchmark/WorkerBenchmark.kt]: runBlocking is here so we can await the enqueue - although it's running on a synchronous executor we still have to await the promise as the API is async.

I'm open to some cleaner / better ideas to write this though as I don't like it much either :smiley:

**Comment by dlam** [comment]
Give me a second to fix some rebase issues :smile:

**Comment by dlam** [comment]
Had to remake the PR, continued here: https://github.com/googlesamples/android-architecture-components/pull/581

Seems to cache outdated commits triggering googlebot - from rebasing ontop of master before updating benchmark-samples confused
